Enhancements

petoem opened the pull request:
See #4

petoem committed:
Enhancements
- expose add_watch and rm_watch
- allow multiple `Proc` to handle events
- change `initialize` to improve reusability
- improve error handling

petoem committed:
Add description to shard.yml

petoem committed:
Various changes
- add Dir#watch and File#watch
- add Inotify.watcher
- disable fallback

petoem committed:
Fix wrong method call

petoem committed:
Add ameba code style linter

petoem committed:
Rename enum EventType to Type

j8r commented:
There are no specs - I can add some in another PR after this one :wink:

j8r commented on shard.yml:
```diff
@@ -1,9 +1,16 @@
 name: inotify
 version: 0.1.0
```
Don't forget to bump to `0.2.0` and do a `git tag v0.2.0 && git push --tags`

j8r commented on shard.yml:
```diff
@@ -1,9 +1,16 @@
+    github: veelenga/ameba
+    version: 0.8.1
+
 crystal: 0.25.1
```
Can be changed to `0.27.0`, or simply removed.

j8r commented on src/inotify/event.cr:
```diff
@@ -2,35 +2,35 @@ module Inotify
+        event = Type::DELETE if LibInotify::IN_DELETE & mask != 0
+        event = Type::DELETE_SELF if LibInotify::IN_DELETE_SELF & mask != 0
+        event = Type::MOVE_SELF if LibInotify::IN_MOVE_SELF & mask != 0
+        event
```
Does the `event` variable needed? Can it be replaced by directly `return`, with `Type::UNKNOWN` at the end?

j8r reviewed (commented):
The `version.cr` file can also be remove - shards doesn't use it.

petoem committed:
Improve Event and Event::Type
- simplify Event `initialize`
- add Event#type_is? method
- mark enum with flags attribute
- add more Event types

petoem commented:
@j8r Yes, having some specs would be great. 

Wish you a happy new year in advance. :fireworks:

petoem committed:
Use `skip_file` macro on other platforms

petoem committed:
Remove inotify fallback

j8r commented on src/inotify/watcher.cr:
```diff
@@ -90,29 +82,50 @@ module Inotify
+      if status == -1
+        case Errno.value
+        when Errno::EBADF then raise IO::Error.new "fd is not a valid file descriptor"
+        when Errno::EINVAL then raise Errno.new "The watch descriptor wd is not valid; or fd is not an inotify file descriptor"
```
It can be:
```cr
case Errno
when .ebadf?  then ...
when .einval? then ...
else          raise ...
end
```

petoem replied:
This won't work. `Errno` is a class not an enum.

j8r replied:
you're right

j8r commented on src/inotify.cr:
```diff
@@ -1,31 +1,47 @@
   end
 end
 
 class File
```
This monkey patching can be easily avoided, there are class methods.
For `#watch`, a method can accept a path or `File`.

petoem replied:
I remove them, they just pollute std-lib `File` and `Dir`.

j8r commented on src/inotify.cr:
```diff
@@ -1,31 +1,47 @@
   end
 end
 
 class Dir
```
Ditto, maybe `Inotify::File` and `Inotify::Dir`?

petoem committed:
Properly stop and resume `Watcher`

petoem committed:
Use macro to set VERSION const

petoem committed:
Don't reopen `File` and `Dir` class

j8r commented on src/inotify/version.cr:
```diff
@@ -1,3 +1,3 @@
 module Inotify
-  VERSION = "0.1.0"
+  VERSION = {{ `shards version #{__DIR__}`.chomp.stringify }}
```
You don't have to bother with this, I suggest to remove this file - does someone user it?

petoem replied:
I am just used to it being there and it doesn't hurt to have it.

petoem committed:
Bump up crystal version

petoem commented:
I just noticed that `@type=None` and not `UNKNOWN` because of the enums flags attribute. :man_facepalming: 
```crystal
Inotify::Event(
 @cookie=0_u32,
 @is_dir=false,
 @mask=32768_u32,
 @name="test.txt",
 @path="./spec/test/test.txt",
 @type=None)
```
`None` conveys a totally different message than `UNKNOWN`.

petoem committed:
Fix Event::Type and minor improvements

petoem commented:
`event.cr` does not need to be platform independent anymore, because I removed inotify fallback. So, I will just assign the `LibInotify` constants to the enum members.

petoem commented:
@j8r I think that's it for now. Specs and docs can be added later in another PR.

Btw, you can use a block with no argument, despite the fact that the `Proc` gets an `Event` passed in.
```crystal
Inotify.watch("./spec/test/a.txt") do
  pp "An event occurred!"
end
```

j8r commented:
I don't really like the `not_nil!` used with `@fd` and `@io`. It's probably due to the enable/disable feature: what is the use case of this, vs. reinstantiating a new object?

petoem commented:
This simplifies temporarily toggling it on/off.
